### PR TITLE
Quick patch on pred removal pass

### PR DESCRIPTION
**Pull request by shmsong**
Quick patch on #1641 for simpler cases to unblock. 

Will need to look further for more thorough fix.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/lower_predicate_elimination.h [hunk 1]
```diff
@@ -26,6 +26,18 @@ class TORCH_CUDA_CU_API PredicateElimination : public IterVisitor {
   //! Dump to string for debugging
   std::string toString() const;
 
+  // A utility to set removal info of `to` the same as `from`.
+  //  See issue #...
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_predicate_elimination.h]: nit: issue number

File: torch/csrc/jit/codegen/cuda/lower_index.cpp [hunk 2]
```diff
@@ -38,6 +38,13 @@ void IndexLowering::pushBack(Expr* expr) {
   }
 }
 
+Expr* IndexLowering::back() const {
+  if (active_scope_ == nullptr) {
+    return lowered_exprs_.back();
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_index.cpp:45]: Can you please add an assertion that `lower_exprs_` and `active_scope_->exprs()` are not empty? I understand they should not become empty, but debugging bugs due to `std::vector::front, std::vector:back` are not something always easy to diagnose.

**Review by naoyam — APPROVED** [review]
Thanks for the fix. Approved as it makes sense. 

I'm however a little concerned about the propagation method as it doesn't seem like a robust way. It would be easy to forget doing so. Also, there may be some other GpuLower maps that also need to be updated. Any thoughts?

**Comment by naoyam** [comment]
Why not add a function like this?

```
void GpuLower::propagateExprInfo(Expr* old_expr, Expr* new_expr) {
  predicate_elimination_.propagateRemovalInfo(old_expr, new_expr);
}
```

And when replacing expressions, instead of directly calling `GpuLower::current()->predicateElimination().propagateRemovalInfo()`, call `GpuLower::current()->propagateExprReplacement()`.
This way, at least, it would be easier to do something more with the GpuLower mapppings in the future.

File: torch/csrc/jit/codegen/cuda/lower_predicate_elimination.h (hunk 1 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/lower_predicate_elimination.h]: updated. Thanks.

File: torch/csrc/jit/codegen/cuda/lower_index.cpp (hunk 2 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/lower_index.cpp:45]: Updated. Thanks.

**Comment by shmsong** [comment]
> Thanks for the fix. Approved as it makes sense.
> 
> I'm however a little concerned about the propagation method as it doesn't seem like a robust way. It would be easy to forget doing so. Also, there may be some other GpuLower maps that also need to be updated. Any thoughts?

Thanks for approving. Yes I agree this is not the way we want to do it long term, and thus `Quick patch ...`. 
We could add improving this pass into the mid-term/short-term plan.

I think this issue could mean that we are starting to see the need for a mechanism to validate/invalidate previously done analysis passes after mutate/transform passes. This shows that we have reached great complexity in our passes and interactions between them, which is really cool 👍 . 

Meanwhile the validation/invalidation is hard to completely automate, since only the transform passes themselves would know if the analysis results are preserved or invalidated. We could invest gradually in some infra to facilitate this.

_For example, say hypothetically in a lowering pass we replaced a welford op into a few pointwise ops, a reduction, and a broadcast (I know this is not the way it's done currently so we don't have this trouble to deal with now.), then there might have been additional predicate removal opportunities in the pointwise ops that were new so the transform pass might as well just ask to re-run predicate removal pass after the replacements._

maybe a candidate solution is, borrowing ideas from other OSS, there could be a pass manager that gives the transform passes an interface to specify which analyses they preserve or invalidate. The invalidated analysis passes will be re-run next time they are pulled by other passes. This will take quite a bit interface refactoring for us to do, so just a thought. 

Only brainstorming ideas at the moment and we could discuss more as the complexity of our lowering continues to grow and see what's the best in our system.

**Comment by naoyam** [comment]
Agreed. The mapping information built up during the lowering becomes quite substantial. Not only validating and updating it, it's also important to ensure the lowering passes are ordered so as to satisfy required mapping information is generated before. Everything is done manually now, which is quite error prone. Each pass should encode which mapping information is required and which information is produced so that the dependency is properly maintained. In any case, not an urgent thing to do, but we should make it happen not in the distant future.

**Pull request merged**